### ISSUE TITLE
[System.Threading.Tasks.Dataflow] Dataflow unreliably completes before processing all items

### ISSUE THREAD

**Issue by Hermholtz** [dotnet/runtime#78772]
### Description

Presence of `Console.WriteLine()` before call to `Complete()` makes dataflow unreliable. Its behavior is not consistent. It sometimes fails to process items.

I've created a simple Dataflow with 2 `TransformBlock`s linked to a single `ActionBlock`. Both links have `PropagateCompletion = true`.

```
input1 (TransformBlock) -----> consumer (ActionBlock)
input2 (TransformBlock) --^
```

I've posted 10 items to both `TransformBlock`s (20 in total) and then called `Complete()` on them. I `await`ed on completion of the `ActionBlock`.

I expected all 20 messages to be processed, but it turns out that it's not always the case. The number of messages is usually 20, but sometimes can be 10. It happens in both Debug and Release modes.

**The bug only occurs when I have a `Console.WriteLine()` before calls to `Complete()`.**

![image](https://user-images.githubusercontent.com/15248889/203574699-3b298f7a-87bc-4724-b348-5c2633723e9c.png)

### Reproduction Steps

- Create a .NET 7 console application in Visual Studio 2022, 
- Paste the code below, build it,
-  Open a `pwsh` window, go to the output directory, for example `C:\DataflowBugTest\bin\Release\net7.0`
- Execute the binary in a loop: `for (;;) { .\DataflowBugTest.exe }`.

```csharp
using System.Threading.Tasks.Dataflow;

int itemsConsumed = 0;

var input1 = new TransformBlock<int, int>(item => 100 + item);
var input2 = new TransformBlock<int, int>(item => 200 + item);
var consumer = new ActionBlock<int>(item => Interlocked.Increment(ref itemsConsumed));

var linkOptions = new DataflowLinkOptions { PropagateCompletion = true };
input1.LinkTo(consumer, linkOptions);
input2.LinkTo(consumer, linkOptions);

for (int i = 0; i < 10; i++)
{
    input1.Post(i);
    input2.Post(i);
}

const int itemsExpected = 20;

Console.WriteLine();  // <--- comment or delete this and the bug will disappear.
input1.Complete();
input2.Complete();

await consumer.Completion;

if (itemsConsumed != itemsExpected)
{
    Console.WriteLine($"ERROR! Dataflow processed {itemsConsumed} items, expected {itemsExpected}.");
}
```

### Expected behavior

In this test code I expect empty lines of output (no message printed).
In Dataflow, I expect a consistent and documented behavior - it should always process all items, and don't lose data.

### Actual behavior

Random occurrences of error message. You may need to wait several seconds or even several minutes before the first one appears.
![image](https://user-images.githubusercontent.com/15248889/203574051-15ea2812-950c-49bd-9097-6157e60d8fb3.png)

### Regression?

I don't know.

### Known Workarounds

For this test code it's enough to delete the `Console.WriteLine()`.
In general - I don't know.

### Configuration

.NET 7.0.0
System.Threading.Tasks.Dataflow 7.0.0 (file version 7.0.22.51805).
Windows 10 10.0.19045.2311 x64
CPU is AMD Ryzen 7 4800H.
Tried also on AMD Ryzen 7 3700U - here I had to wait several minutes, but it also happened.
No other hardware available to test at the moment.

.NET SDK:
 Version:   7.0.100
 Commit:    e12b7af219

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.100\

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

**Comment by svick** [comment]
> I've created a simple Dataflow with 2 `TransformBlock`s linked to a single `ActionBlock`. Both links have `PropagateCompletion = true`.

This means that the `ActionBlock` completes as soon as *one* of the `TransformBlock`s completes. Which means Dataflow behaving non-deterministically is correct.

**Comment by Hermholtz** [comment]
> his means that the `ActionBlock` completes as soon as _one_ of the `TransformBlock`s completes. So Dataflow behaving non-deterministically here is correct.

Sorry but where it is documented?

**Comment by svick** [comment]
> Sorry but where it is documented?

[The documentation for `PropagateCompletion` says:](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.dataflowlinkoptions.propagatecompletion?view=net-7.0)

> Gets or sets whether the linked target will have completion and faulting notification propagated to it automatically.

That's not quite clear, but any one source block completing a linked target block is definitely a valid interpretation.

There's also [*Walkthrough: Creating a Dataflow Pipeline*](https://learn.microsoft.com/en-us/dotnet/standard/parallel-programming/walkthrough-creating-a-dataflow-pipeline), which says:

> If you also provide `DataflowLinkOptions` with `PropagateCompletion` set to true, successful or unsuccessful completion of one block in the pipeline will cause completion of the next block in the pipeline.

Though that's still not completely clear and I think the documentation could be improved.

**Comment by Hermholtz** [comment]
Thank you. So in this case, what's the proper way to terminate the pipeline such as the above?

I'd like a few threads/tasks to pump data, each into its own `TransformBlock`, because of the transformation that are done, and have a method of signalling from that thread/task that it has finished, and only when all of the threads/tasks has finished delivering data, signal the `ActionBlock` that there won't be more data and processing is finished. 

What would be the flow of the program then, instead of intuitive `await consumer.Complete`? I thought it's "AND" (it completes only if ALL of its links have completed), but from your description it's OR (complete if ANY of links have completed).

**Comment by svick** [comment]
> So in this case, what's the proper way to terminate the pipeline such as the above?

You can propagate completion manually, e.g. using [this code from SO](https://stackoverflow.com/a/61639105/41071):

```c#
public static async void PropagateCompletion(IDataflowBlock[] sources, IDataflowBlock target)
{
    Task allSourcesCompletion = Task.WhenAll(sources.Select(s => s.Completion));

    try { await allSourcesCompletion.ConfigureAwait(false); } catch { }

    var exception = allSourcesCompletion.IsFaulted ? allSourcesCompletion.Exception : null;

    if (exception != null)
        target.Fault(exception);
    else
        target.Complete();
}
```

**Comment by Hermholtz** [comment]
Thank you for this code snippet. It'd work in my simple case.

I think this opens a can of worms. The lack of automatic completion propagation in the "AND" fashion makes Dataflow very difficult to use. It works fine today in sequential flows, such as TransformBlock->BufferBlock->ActionBlock, but when there are multiple sources linked to a single target, and more branches, using more than just a few blocks, then it's a design failure. It requires a verbose and explicit completion propagation code, such as the above snippet, whereas a simple feature could have make it much, much easier to use.

The feature would be an option that would tell a block to set its completion status _if and only if_ **all** its linked sources are completed (and as soon as it happens). 

This would enable creating sophisticated flows/DAGs made of more than a couple of blocks that would correctly track completion and allow pipeline/program termination easily at the right moment, by using just a single `await terminalBlock.Completion`.

Otherwise it's not worth using Dataflow at all, the additional code complexity can be applied directly to Channels, ConcurrentQueues or other thread-safe buffers. Which is a pity, because Dataflow has great potential of creating scalable, high performance, complex data processing pipelines made of reusable blocks in a very easy way.

**Comment by svick** [comment]
> The feature would be an option that would tell a block to set its completion status _if and only if_ **all** its linked sources are completed (and as soon as it happens).

As I understand it, that wouldn't work, because the target block does not know who its sources are, or how many of them are there. (And it's probably too late to change that.)

Also, it wouldn't work well if blocks are being linked while other blocks are already completing.

**Comment by Hermholtz** [comment]
> the target block does not know who its sources are, or how many of them are there. (And it's probably too late to change that.)

That looks like an easy one to implement. There could be a private list of source blocks and their completion status (in SourceCore or TargetCore maybe), filled by the `LinkTo` method, for example:
```
internal class LinkedSourceBlockItem
{
    internal IReceivableSourceBlock<T> LinkedSourceBlock;  // accessibility and design details TBD
    internal bool IsCompleted;
}
```

The more difficult thing I see is how to get the information about which source block has called the `IDataflowBlock.Complete()` method. There's no parameter to the `Complete()` method today (now I understand the semantic - whoever calls it, it's called, period...). But maybe there _is_ a way to know that without making a breaking change to the IDataflowBlock interface.

> Also, it wouldn't work well if blocks are being linked while other blocks are already completing.

Thinking loudly, it seems that if `LinkTo` encounters a source block which is already Completed (maybe it could inspect it?) then just set that above `IsCompleted` bit and be done. On the other hand, if our (target) block is already completed, and a source block is being linked to it, it doesn't change our state. It's the same today - if I link another source block to a completed block, it doesn't reset it completion status.

Crucial part is that every time the `IsCompleted` is set on any of the above list members, there should be code that checks if it is set for every linked blocks on the list, and if so then set this block's state to Completed and trigger appropriate actions.

Hope it makes sense.

**Comment by stephentoub** [comment]
> You can propagate completion manually, e.g. using [this code from SO](https://stackoverflow.com/a/61639105/41071):

Yes, or just:
```C#
public static void PropagateCompletion(IDataflowBlock[] sources, IDataflowBlock target) =>
    Task.WhenAll(sources.Select(s => s.Completion)).ContinueWith(t =>
    {
        if (t.IsFaulted) target.Fault(t.Exception);
        else target.Complete();
    }
}
```

> I think this opens a can of worms. The lack of automatic completion propagation in the "AND" fashion makes Dataflow very difficult to use.

It is by design that any such "all" behavior is left up to the constructor of the graph, and developers have been successfully using the dataflow library for over a decade with that.  It would be a significant challenge and expense to add a "complete when all sources to a block complete", as a) that information isn't communicated or tracked today, b) the structure of a graph can be dynamic, with blocks linking and unlinking at will, and c) the API itself is not conducive to an arbitrarily flexible notion of joint completion (e.g. at the point where you say you want "all" behavior, you also need to consider partial behavior, such as "I want to flow my completion but only when these other two specific blocks have also completed but we don't care about those other three blocks", which is something easily achieved with a PropagateCompletion helper like that above but much harder trying to somehow build that concept into the blocks themselves).  Long ago we decided it was best to leave it up to the constructor of the graph exactly how and when completion flowed, the only special-case to that being the super simple 1:1 case, which is also the most common case. I do not see us changing that now, but thank you for the suggestion; if this means the dataflow library isn't appropriate for your needs, then you can obviously build and/or use something else.  And if you decide to keep using the dataflow library and you come up with what you believe is a good model for this "all" completion that you can layer on top, I'd encourage you to put it into a reusable library on nuget for others to consume and vote with their feet.

Thanks.

**Comment by Hermholtz** [comment]
@stephentoub Thank you for your elaborate explanation. It's really informative.

> this means the dataflow library isn't appropriate for your needs

I really like the concept of components which do a well defined stuff with data, and creation of DAGs or Directed Acyclic Graphs (or _networks_, as you call it) out of them. Think Azure Data Factory, SQL Server Integration Services, or Apache Beam that abstracts away low-level infrastructural machinery. I just maybe misunderstand some of the concepts of the .NET Dataflow library. To me it's natural to expect that the flow network only completes if all data items have been safely processed, and to want to have as little infra boilerplate code as possible, because that can be source of subtle bugs.

It would be nice to have more information on Dataflow. I'm aware of this [Dataflow](https://learn.microsoft.com/en-us/dotnet/standard/parallel-programming/dataflow-task-parallel-library) chapter in the documentation but it doesn't give me the answers to the issues below.

First to improve would be the documentation of classes, which is seriously lacking, having just sparse sentence here or there. Look for example at [TransformBlock<TInput,TOutput> Class](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.transformblock-2). Not a single line of code or description or reference to walkthroughs, not to mention missing more advanced information. You'll ask what kind of information? Even [this guidance](https://learn.microsoft.com/en-us/dotnet/standard/parallel-programming/dataflow-task-parallel-library#transformblocktinput-toutput) doesn't inform that TransformBlock contains **a buffer**. Can I rely on that buffer, or do I need a separate, additional BufferBlock before/after TransformBlock? I mean can I just throw a million data items at TransformBlock when its transformation is kinda compute intensive? Will it buffer, or will it lose data? I don't want to learn it hard in production (such as the original issue which I reported above).

And when it comes to orchestration and propagation of completion, there's not much of a concrete example as well. How am I supposed to use the code you provided within this example DAG below? How often and where to call it? Where should I store the arrays of sources?

No offence, I'm just lost, really wanting to use Dataflow in new architectures, just need it to solve problems, not create more.

Thanks in advance for providing answers.

![image](https://user-images.githubusercontent.com/15248889/204238476-70e6befc-3f00-427d-a9e5-6c204125127d.png)

**Comment by stephentoub** [comment]
Given a helper like the previously cited:
```C#
public static void PropagateCompletion(IDataflowBlock[] sources, IDataflowBlock target);
```
with your graph you shared above, you'd add:
```C#
PropagateCompletion(new IDataflowBlock[] { Input1, Input2 }, Buffer1);
PropagateCompletion(new IDataflowBlock[] { Input3, Input4 }, Buffer2);
PropagateCompletion(new IDataflowBlock[] { Buffer1, Buffer2 }, Action1);
```

**Comment by Hermholtz** [comment]
For anybody who may be looking for complete example - this is how to handle completion so that this issue doesn't occur.
The explanation should be outlined in Dataflow documentation in multiple places. The default behavior is not intuitive, unexpected and in more complex networks of blocks can cause silent data loss.

This is a .NET 7 Console Application, should also work in a few earlier versions. Here's the fiddle: https://dotnetfiddle.net/opt5O8

```csharp
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using System.Threading.Tasks.Dataflow;

int itemsConsumed = 0;

var input1 = new TransformBlock<int, int>(item => item);
var input2 = new TransformBlock<int, int>(item => item);
var input3 = new TransformBlock<int, int>(item => item);
var input4 = new TransformBlock<int, int>(item => item);

var buffer1 = new BufferBlock<int>();
var buffer2 = new BufferBlock<int>();

var action1 = new ActionBlock<int>(item =>
    Interlocked.Increment(ref itemsConsumed));

input1.LinkTo(buffer1);
input2.LinkTo(buffer1);

input3.LinkTo(buffer2);
input4.LinkTo(buffer2);

buffer1.LinkTo(action1);
buffer2.LinkTo(action1);

// You have to write completion propagation rules explicitly yourself.
// buffer1 will complete only after both inputs will have completed.
PropagateCompletion(new[] { input1, input2 }, buffer1);
PropagateCompletion(new[] { input3, input4 }, buffer2);
PropagateCompletion(new[] { buffer1, buffer2 }, action1);

const int items = 100000;

for (int i = 0; i < items; i++)
{
    input1.Post(1*items + i);
    input2.Post(2*items + i);
    input3.Post(3*items + i);
    input4.Post(4*items + i);
}

input1.Complete();
input2.Complete();
input3.Complete();
input4.Complete();

await action1.Completion;

Console.WriteLine($"Dataflow processed {itemsConsumed} items.");

// This is the helper method suggested here.
static void PropagateCompletion(IDataflowBlock[] sources, IDataflowBlock target) =>
    Task.WhenAll(sources.Select(s => s.Completion)).ContinueWith(t =>
    {
        if (t.IsFaulted)
            target.Fault(t.Exception);
        else
            target.Complete();
    });
```

**Comment by stephentoub** [comment]
Closing as answered. Thanks.
